From f02c786f59b5082fbf92f03d45de2bfc687e9c43 Mon Sep 17 00:00:00 2001
From: Gu Chao <gux.chao@intel.com>
Date: Fri, 6 Mar 2020 14:21:47 +0800
Subject: [PATCH] DRV_DPM_SW-179: DPM cleanup klocwork issues

- fix various signed unsigned value comparision issues
- fix get_dma_flags_str31() with larger string buffer size and init
  flags varaible properly
- convert nested if-else to switch in various places
- fix get_offset_clear_chksum()
- refactor get_base_dev()
- fix dp_platform_queue_set() with range check on cqm deq port
- fix dp_swdev_port_obj_del()
- remove unreachable code in proc_port_dump()
- change return type of dp_map_to_drop_q() to void
- change return type of dp_platform_color_table_set() to void
- add DP_NODE_SET_CMD_MAX to enum dp_node_en to avoid klocwork warning
- remove unused error cases in dp_xmit_31()
---
 drivers/net/datapath/dpm/datapath.h                | 39 +++++--------
 drivers/net/datapath/dpm/datapath_api.c            | 47 +++++++---------
 drivers/net/datapath/dpm/datapath_logical_dev.c    | 35 +++++-------
 drivers/net/datapath/dpm/datapath_misc.c           |  4 +-
 drivers/net/datapath/dpm/datapath_proc.c           |  5 --
 drivers/net/datapath/dpm/datapath_proc_qos.c       |  4 +-
 drivers/net/datapath/dpm/datapath_swdev.c          | 16 ++----
 drivers/net/datapath/dpm/gswip31/datapath_gswip.c  | 65 +++++++++++++++-------
 drivers/net/datapath/dpm/gswip31/datapath_misc.c   | 12 ++--
 drivers/net/datapath/dpm/gswip31/datapath_ppv4.h   |  2 +-
 .../net/datapath/dpm/gswip31/datapath_ppv4_api.c   | 61 ++++++++++----------
 drivers/net/datapath/dpm/gswip31/datapath_proc.c   |  3 +-
 .../net/datapath/dpm/gswip31/datapath_switchdev.c  |  2 +-
 drivers/net/datapath/dpm/gswip31/datapath_tx.c     | 25 ---------
 include/net/datapath_api_qos.h                     |  3 +-
 15 files changed, 141 insertions(+), 182 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index d322865724dd..032b27ce25e4 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -95,26 +95,22 @@
 #define DP_DEBUG(flags, fmt, arg...)
 #endif				/* end of CONFIG_INTEL_DATAPATH_DBG */
 
-static inline int is_invalid_port(u32 port_id)
+static inline bool is_invalid_port(unsigned int port_id)
 {
-	if (port_id >= MAX_DP_PORTS || port_id < 0) {
-		pr_err("port_id(%d) should be with 0 - %d\n", port_id,
-		       MAX_DP_PORTS);
-		return -1;
-	} else {
-		return 0;
+	if (port_id >= MAX_DP_PORTS) {
+		pr_err("port_id(%u) should be 0~%u\n", port_id, MAX_DP_PORTS);
+		return true;
 	}
+	return false;
 }
 
-static inline int is_invalid_inst(u32 inst)
+static inline bool is_invalid_inst(unsigned int inst)
 {
-	if (inst < 0 || inst >= DP_MAX_INST) {
-		pr_err("inst(%d) should be with 0 - %d\n", inst,
-		       DP_MAX_INST);
-		return -1;
-	} else {
-		return 0;
+	if (inst >= DP_MAX_INST) {
+		pr_err("inst(%u) should be 0~%u\n", inst, DP_MAX_INST);
+		return true;
 	}
+	return false;
 }
 
 #define IFNAMSIZ 16
@@ -829,16 +825,12 @@ extern struct dma_chan_info *dp_dma_chan_tbl[DP_MAX_INST];
 
 static inline struct inst_property *get_dp_port_prop(int inst)
 {
-	if (inst >= 0)
-		return &dp_port_prop[inst];
-	return NULL;
+	return &dp_port_prop[inst];
 }
 
 static inline struct pmac_port_info *get_dp_port_info(int inst, int index)
 {
-	if (index < dp_port_prop[inst].info.cap.max_num_dp_ports)
-		return &dp_port_info[inst][index];
-	return NULL;
+	return &dp_port_info[inst][index];
 }
 
 static inline struct cqm_port_info *get_dp_deqport_info(int inst, int idx)
@@ -858,9 +850,7 @@ static inline struct q_info *get_dp_q_info(int inst, int idx)
 
 static inline struct inst_info *get_dp_prop_info(int inst)
 {
-	if (inst >= 0)
-		return &dp_port_prop[inst].info;
-	return NULL;
+	return &dp_port_prop[inst].info;
 }
 
 struct inst_property *dp_port_prop_dbg(int inst);
@@ -970,7 +960,6 @@ int get_offset_clear_chksum(struct sk_buff *skb, u32 *ip_offset,
 int get_vlan_info(struct net_device *dev, struct vlan_info *vinfo);
 int dp_basic_proc(void);
 
-struct inst_property *get_dp_port_prop(int inst);
 struct pmac_port_info *get_port_info_via_dp_port(int inst, int dp_port);
 
 void set_dp_dbg_flag(uint32_t flags);
@@ -1035,7 +1024,7 @@ struct dp_subif_cache *dp_subif_lookup_safe(struct hlist_head *head,
 int dp_subif_list_init(void);
 int parser_enabled(int ep, struct dma_rx_desc_1 *desc_1);
 int dp_lan_wan_bridging(int port_id, struct sk_buff *skb);
-u32 get_dma_chan_idx(int inst, int num_dma_chan);
+int get_dma_chan_idx(int inst, int num_dma_chan);
 u32 alloc_dma_chan_tbl(int inst);
 void free_dma_chan_tbl(int inst);
 u32 alloc_dp_port_subif_info(int inst);
diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index adecebe1cb70..cb09f22a57af 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -448,7 +448,7 @@ static int32_t dp_alloc_port_private(int inst,
 	port->policy_num = 4;
 #endif
 	if (cbm_data.num_dma_chan) {
-		u16 dma_ch_base;
+		int dma_ch_base;
 
 		dma_ch_base = get_dma_chan_idx(inst, cbm_data.num_dma_chan);
 		if (dma_ch_base == DP_FAILURE) {
@@ -456,8 +456,7 @@ static int32_t dp_alloc_port_private(int inst,
 			cbm_dp_port_dealloc(owner, dev_port, port_id, &cbm_data,
 					    flags | DP_F_DEREGISTER);
 			/* Only clear those fields we need to clear */
-			memset(port, 0,
-			       offsetof(struct pmac_port_info, tail));
+			memset(port, 0, offsetof(struct pmac_port_info, tail));
 			return DP_FAILURE;
 		}
 		port->dma_ch_base = dma_ch_base;
@@ -633,18 +632,13 @@ int32_t dp_register_subif_private(int inst, struct module *owner,
 		/*now find a free subif or valid subif
 		 *need to do configuration HW
 		 */
-		if (port_info->status) {
-			if (dp_port_prop[inst].info.subif_platform_set(
-				    inst, port_id, i, &platfrm_data, flags)) {
-				pr_err("subif_platform_set fail\n");
-				goto EXIT;
-			} else {
-				DP_DEBUG(DP_DBG_FLAG_REG,
-					 "subif_platform_set succeed\n");
-			}
+		if (dp_port_prop[inst].info.subif_platform_set(
+			inst, port_id, i, &platfrm_data, flags)) {
+			pr_err("subif_platform_set fail\n");
+			goto EXIT;
 		} else {
-			pr_err("port info status fail for 0\n");
-			return res;
+			DP_DEBUG(DP_DBG_FLAG_REG,
+				 "subif_platform_set succeed\n");
 		}
 		sif->flags = 1;
 		sif->netif = dev;
@@ -1296,15 +1290,9 @@ int32_t dp_register_dev_ext(int inst, struct module *owner, u32 port_id,
 
 	DP_LIB_LOCK(&dp_lock);
 	if (flags & DP_F_DEREGISTER) {	/*de-register */
-		if (port_info->status != PORT_DEV_REGISTERED) {
-			DP_DEBUG(DP_DBG_FLAG_REG,
-				 "No or %s to de-register for num_subif=%d\n",
-				 owner->name, port_info->num_subif);
-		} else if (port_info->status == PORT_DEV_REGISTERED) {
-
+		if (port_info->status == PORT_DEV_REGISTERED) {
 			dp_notifier_invoke(inst, port_info->dev, port_id,
 					   DP_EVENT_DE_REGISTER_DEV);
-
 			port_info->status = PORT_ALLOCATED;
 			DP_CB(inst, dev_platform_set)(inst, port_id, data,
 						      flags);
@@ -1314,10 +1302,9 @@ int32_t dp_register_dev_ext(int inst, struct module *owner, u32 port_id,
 			res = DP_SUCCESS;
 		} else {
 			DP_DEBUG(DP_DBG_FLAG_REG,
-				 "No for %s to de-register for unknown status:%d\n",
-				 owner->name, port_info->status);
+				 "fail to de-register for unknown status:%d\n",
+				 port_info->status);
 		}
-
 		DP_LIB_UNLOCK(&dp_lock);
 		return res;
 	}
@@ -2217,15 +2204,19 @@ int dp_set_pmapper(struct net_device *dev, struct dp_pmapper *mapper, u32 flag)
 	if (!map)
 		return DP_FAILURE;
 	memcpy(map, mapper, sizeof(*map));
-	if (mapper->mode == DP_PMAP_PCP || mapper->mode == DP_PMAP_DSCP) {
+
+	switch (mapper->mode) {
+	case DP_PMAP_PCP:
+	case DP_PMAP_DSCP:
 		map->mode = GSW_PMAPPER_MAPPING_PCP;
-	} else if (mapper->mode == DP_PMAP_DSCP_ONLY) {
+		break;
+	case DP_PMAP_DSCP_ONLY:
 		map->mode = GSW_PMAPPER_MAPPING_DSCP;
-	} else {
+		break;
+	default:
 		pr_err("Unknown mapper mode: %d\n", map->mode);
 		goto EXIT;
 	}
-
 	/* workaround in case caller forget to set to default ctp */
 	if (mapper->mode == DP_PMAP_PCP)
 		for (i = 0; i < DP_PMAP_DSCP_NUM; i++)
diff --git a/drivers/net/datapath/dpm/datapath_logical_dev.c b/drivers/net/datapath/dpm/datapath_logical_dev.c
index 04a787ada418..57d0145b7747 100644
--- a/drivers/net/datapath/dpm/datapath_logical_dev.c
+++ b/drivers/net/datapath/dpm/datapath_logical_dev.c
@@ -73,32 +73,27 @@ struct logic_dev *logic_list_lookup(struct list_head *head,
 	return NULL;
 }
 
-/*level >=1: only get the specified level if possible
- *otherwise: -1 maximum
+/* level > 0: search until specified level is reached
+ * level < 0: search for maximum depth
  */
 struct net_device *get_base_dev(struct net_device *dev, int level)
 {
-	struct net_device *lower_dev, *tmp;
+	struct net_device *cdev = dev;
+	struct net_device *ldev;
 	struct list_head *iter;
 
-	tmp = dev;
-	lower_dev = NULL;
-	do {
-		netdev_for_each_lower_dev(tmp, lower_dev, iter)
-		break;
-		if (lower_dev) {
-			tmp = lower_dev;
-			lower_dev = NULL;
-			level--;
-			if (level == 0)
-				break;
-		} else {
+	/* klocwork workaround
+	 * code logic comes from netdev_for_each_lower_dev
+	 */
+	while (cdev && level) {
+		iter = cdev->adj_list.lower.next,
+		ldev = netdev_lower_get_next(cdev, &iter);
+		if (!ldev)
 			break;
-		}
-	} while (1);
-	if (tmp == dev)
-		return NULL;
-	return tmp;
+		cdev = ldev;
+		level--;
+	}
+	return cdev == dev ? NULL : cdev;
 }
 
 /* add logic device into its base dev's logic dev list */
diff --git a/drivers/net/datapath/dpm/datapath_misc.c b/drivers/net/datapath/dpm/datapath_misc.c
index 1ba13a88c76a..4af020c5c629 100644
--- a/drivers/net/datapath/dpm/datapath_misc.c
+++ b/drivers/net/datapath/dpm/datapath_misc.c
@@ -388,7 +388,7 @@ int get_offset_clear_chksum(struct sk_buff *skb, u32 *ip_offset,
 {
 	u8 *p_l2_mac = skb->data;
 	u8 *p = p_l2_mac + TWO_MAC_SIZE;
-	struct ip_hdr_info pkt_info[2];
+	struct ip_hdr_info pkt_info[3];
 	u8 ip_num = 0;
 #if IS_ENABLED(CONFIG_INTEL_DATAPATH_DBG)
 	int i;
@@ -1116,7 +1116,7 @@ int dp_cpufreq_notify_exit(void)
  * Description: Find free dma channel index from dp_dma_chan_tbl.
  * Return: Base idx on success DP_FAILURE on failure.
  */
-u32 get_dma_chan_idx(int inst, int num_dma_chan)
+int get_dma_chan_idx(int inst, int num_dma_chan)
 {
 	u32 base, match;
 
diff --git a/drivers/net/datapath/dpm/datapath_proc.c b/drivers/net/datapath/dpm/datapath_proc.c
index 463352032230..1f618be50203 100644
--- a/drivers/net/datapath/dpm/datapath_proc.c
+++ b/drivers/net/datapath/dpm/datapath_proc.c
@@ -233,11 +233,6 @@ int proc_port_dump(struct seq_file *s, int pos)
 			seq_printf(s, "Reserved Port: %s=0x%08x  %s=0x%08x\n",
 				   "rx_err_drop", STATS_GET(port->rx_err_drop),
 				   "tx_err_drop", STATS_GET(port->tx_err_drop));
-			if (print_ctp_bp && pos) /* just to print bridge
-						  * port zero's member.
-						  * CPU port no ctp/bridge port
-						  */
-				print_ctp_bp(s, tmp_inst, port, 0, 0);
 			seq_printf(s, "           qid/node:     %d/%d\n",
 				   sif->qid, sif->q_node[0]);
 			seq_printf(s, "           port/node:    %d/%d\n",
diff --git a/drivers/net/datapath/dpm/datapath_proc_qos.c b/drivers/net/datapath/dpm/datapath_proc_qos.c
index 83ed5c759782..2eb106721139 100644
--- a/drivers/net/datapath/dpm/datapath_proc_qos.c
+++ b/drivers/net/datapath/dpm/datapath_proc_qos.c
@@ -72,7 +72,7 @@ static char *get_node_stat(int node_id, int type);
 char *get_node_pri(int node_id, int type);
 char *dp_port_flag_str(int cqm_deq_port, int flag);
 char *dp_port_dma_tx_str(int cqm_deq_port, int flag);
-static void conv_limit_to_str(int shaper_limit, char *buf, int size);
+static void conv_limit_to_str(u32 shaper_limit, char *buf, int size);
 
 struct box_info *find_box_via_nodeid(struct box_info *box,
 				     int box_num, int sch_node_id)
@@ -792,7 +792,7 @@ char *get_node_pri(int node_id, int type)
 	return stat;
 }
 
-void conv_limit_to_str(int shaper_limit, char *buf, int size)
+void conv_limit_to_str(u32 shaper_limit, char *buf, int size)
 {
 	if (shaper_limit == DP_NO_SHAPER_LIMIT)
 		snprintf(buf, size, "NoLim");
diff --git a/drivers/net/datapath/dpm/datapath_swdev.c b/drivers/net/datapath/dpm/datapath_swdev.c
index edf581dd9eb1..5fc3e618ca7f 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.c
+++ b/drivers/net/datapath/dpm/datapath_swdev.c
@@ -156,14 +156,12 @@ int dp_swdev_bridge_id_entry_free(int instance)
 {
 	u32 index;
 	struct hlist_node *tmp;
+	struct hlist_head *head;
 	struct br_info *br_item;
 
-	for (index = 0; index < BR_ID_ENTRY_HASH_TABLE_SIZE;
-		index++) {
-		hlist_for_each_entry_safe
-			(br_item, tmp,
-			 &g_bridge_id_entry_hash_table[instance][index],
-			 br_hlist) {
+	for (index = 0; index < BR_ID_ENTRY_HASH_TABLE_SIZE; index++) {
+		head = &g_bridge_id_entry_hash_table[instance][index];
+		hlist_for_each_entry_safe(br_item, tmp, head, br_hlist) {
 			dp_swdev_remove_bridge_id_entry(br_item);
 		}
 	}
@@ -890,8 +888,6 @@ static int dp_swdev_port_attr_get(struct net_device *dev,
 		br_info = dp_swdev_bridge_entry_lookup(br_dev->name);
 		if (!br_info)
 			return 0;
-		if (br_info->fid < 0)
-			return -EOPNOTSUPP;
 		attr->u.ppid.id_len = sizeof(br_info->fid);
 		memcpy(&attr->u.ppid.id, &br_info->fid, attr->u.ppid.id_len);
 		DP_DEBUG(DP_DBG_FLAG_SWDEV,
@@ -938,15 +934,13 @@ static int dp_swdev_port_obj_del(struct net_device *dev,
 				 const struct switchdev_obj *obj)
 {
 	int err = -EOPNOTSUPP;
-	return err;//TODO
 
 	DP_DEBUG(DP_DBG_FLAG_SWDEV, "%s\r\n", __func__);
 	switch (obj->id) {
 	case SWITCHDEV_OBJ_ID_PORT_VLAN:
 		break;
 	default:
-		err = -EOPNOTSUPP;
-		break;
+		return -EOPNOTSUPP;
 	}
 	return err;
 }
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_gswip.c b/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
index c56fb9940edf..f55063f836ad 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
@@ -627,7 +627,7 @@ int alloc_bridge_port(int inst, int port_id, int subif_ix,
 	/*allocate a free bridge port */
 	memset(&bp, 0, sizeof(bp));
 	ret = gsw_bp->BridgePort_Alloc(gsw_handle, &bp);
-	if (ret != GSW_statusOk || bp.nBridgePortId < 0) {
+	if (ret != GSW_statusOk) {
 		pr_err("Failed to get a bridge port\n");
 		return -1;
 	}
@@ -991,7 +991,7 @@ int dp_meter_alloc_31(int inst, int *meterid, int flag)
 	}
 	memset(&meter_cfg, 0, sizeof(meter_cfg));
 	ret = gsw_qos->QOS_MeterAlloc(gsw_handle, &meter_cfg);
-	if (ret != GSW_statusOk || meter_cfg.nMeterId < 0) {
+	if (ret != GSW_statusOk) {
 		pr_err("Failed to get a meter alloc\n");
 		*meterid = -1;
 		return -1;
@@ -1130,14 +1130,19 @@ int dp_meter_add_31(struct net_device *dev,  struct dp_meter_cfg  *meter,
 	meter_cfg.nEbs = meter->pbs;
 	meter_cfg.nPiRate = METER_PIR(meter->pir);
 	meter_cfg.bEnable = 1;
-	if (meter->type == srTCM) {
+
+	switch(meter->type) {
+	case srTCM:
 		meter_cfg.eMtrType = GSW_QOS_Meter_srTCM;
-	} else if (meter->type == trTCM) {
+		break;
+	case trTCM:
 		meter_cfg.eMtrType = GSW_QOS_Meter_trTCM;
-	} else {
+		break;
+	default:
 		pr_err(" invalid meter type\n");
 		return -1;
 	}
+
 	meter_cfg.nMeterId = meter->meter_id;
 	meter_cfg.nColourBlindMode = meter->col_mode;
 	ret = gsw_qos->QoS_MeterCfgSet(gsw_handle, &meter_cfg);
@@ -1233,22 +1238,27 @@ int dp_meter_add_31(struct net_device *dev,  struct dp_meter_cfg  *meter,
 		if (meter->dir == DP_DIR_EGRESS) {
 			bp_cfg->eMask =
 				GSW_BRIDGE_PORT_CONFIG_MASK_EGRESS_SUB_METER;
-			if (meter->dp_pce.flow == DP_UKNOWN_UNICAST)
+			switch(meter->dp_pce.flow) {
+			case DP_UKNOWN_UNICAST:
 				BP_CFG(bp_cfg,
 				       GSW_BRIDGE_PORT_EGRESS_METER_UNKNOWN_UC,
 				       meter->meter_id, 1);
-			else if (meter->dp_pce.flow == DP_MULTICAST)
+				break;
+			case DP_MULTICAST:
 				BP_CFG(bp_cfg,
 				       GSW_BRIDGE_PORT_EGRESS_METER_MULTICAST,
 				       meter->meter_id, 1);
-			else if (meter->dp_pce.flow == DP_BROADCAST)
+				break;
+			case DP_BROADCAST:
 				BP_CFG(bp_cfg,
 				       GSW_BRIDGE_PORT_EGRESS_METER_BROADCAST,
 				       meter->meter_id, 1);
-			else
+				break;
+			default:
 				BP_CFG(bp_cfg,
 				       GSW_BRIDGE_PORT_EGRESS_METER_OTHERS,
 				       meter->meter_id, 1);
+			}
 		} else if (meter->dir == DP_DIR_INGRESS) {
 			bp_cfg->eMask =
 				GSW_BRIDGE_PORT_CONFIG_MASK_INGRESS_METER;
@@ -1278,19 +1288,23 @@ int dp_meter_add_31(struct net_device *dev,  struct dp_meter_cfg  *meter,
 			goto err;
 		}
 		br_cfg->eMask = GSW_BRIDGE_CONFIG_MASK_SUB_METER;
-		if (meter->dp_pce.flow == DP_UKNOWN_UNICAST)
+		switch(meter->dp_pce.flow) {
+		case DP_UKNOWN_UNICAST:
 			BR_CFG(br_cfg,
 			       GSW_BRIDGE_PORT_EGRESS_METER_UNKNOWN_UC,
 			       meter->meter_id, 1);
-		else if (meter->dp_pce.flow == DP_MULTICAST)
+			break;
+		case DP_MULTICAST:
 			BR_CFG(br_cfg,
 			       GSW_BRIDGE_PORT_EGRESS_METER_MULTICAST,
 			       meter->meter_id, 1);
-		else if (meter->dp_pce.flow == DP_BROADCAST)
+			break;
+		case DP_BROADCAST:
 			BR_CFG(br_cfg,
 			       GSW_BRIDGE_PORT_EGRESS_METER_BROADCAST,
 			       meter->meter_id, 1);
-		else {
+			break;
+		default:
 			pr_err("Meter flow invalid\n");
 			goto err;
 		}
@@ -1417,22 +1431,27 @@ int dp_meter_del_31(struct net_device *dev,  struct dp_meter_cfg  *meter,
 		if (meter->dir == DP_DIR_EGRESS) {
 			bp_cfg->eMask =
 				GSW_BRIDGE_PORT_CONFIG_MASK_EGRESS_SUB_METER;
-			if (meter->dp_pce.flow == DP_UKNOWN_UNICAST)
+			switch(meter->dp_pce.flow) {
+			case DP_UKNOWN_UNICAST:
 				BP_CFG(bp_cfg,
 				       GSW_BRIDGE_PORT_EGRESS_METER_UNKNOWN_UC,
 				       meter->meter_id, 0);
-			else if (meter->dp_pce.flow == DP_MULTICAST)
+				break;
+			case DP_MULTICAST:
 				BP_CFG(bp_cfg,
 				       GSW_BRIDGE_PORT_EGRESS_METER_MULTICAST,
 				       meter->meter_id, 0);
-			else if (meter->dp_pce.flow == DP_BROADCAST)
+				break;
+			case DP_BROADCAST:
 				BP_CFG(bp_cfg,
 				       GSW_BRIDGE_PORT_EGRESS_METER_BROADCAST,
 				       meter->meter_id, 0);
-			else
+				break;
+			default:
 				BP_CFG(bp_cfg,
 				       GSW_BRIDGE_PORT_EGRESS_METER_OTHERS,
 				       meter->meter_id, 0);
+			}
 		} else {
 			bp_cfg->eMask =
 				GSW_BRIDGE_PORT_CONFIG_MASK_INGRESS_METER;
@@ -1459,19 +1478,23 @@ int dp_meter_del_31(struct net_device *dev,  struct dp_meter_cfg  *meter,
 			goto err;
 		}
 		br_cfg->eMask = GSW_BRIDGE_CONFIG_MASK_SUB_METER;
-		if (meter->dp_pce.flow == DP_UKNOWN_UNICAST)
+		switch(meter->dp_pce.flow) {
+		case DP_UKNOWN_UNICAST:
 			BR_CFG(br_cfg,
 			       GSW_BRIDGE_PORT_EGRESS_METER_UNKNOWN_UC,
 			       meter->meter_id, 0);
-		else if (meter->dp_pce.flow == DP_MULTICAST)
+			break;
+		case DP_MULTICAST:
 			BR_CFG(br_cfg,
 			       GSW_BRIDGE_PORT_EGRESS_METER_MULTICAST,
 			       meter->meter_id, 0);
-		else if (meter->dp_pce.flow == DP_BROADCAST)
+			break;
+		case DP_BROADCAST:
 			BR_CFG(br_cfg,
 			       GSW_BRIDGE_PORT_EGRESS_METER_BROADCAST,
 			       meter->meter_id, 0);
-		else {
+			break;
+		default:
 			pr_err("Meter flow invalid\n");
 			goto err;
 		}
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index 673e6cc7fcee..f6d67bc8ecb0 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -979,7 +979,7 @@ static int dp_gswip_color_8P0D_set(int mode, int inst)
 	return 0;
 }
 
-static int dp_platform_color_table_set(int inst)
+static void dp_platform_color_table_set(int inst)
 {
 	dp_gswip_color_8P0D_set(GSW_MARKING_PCP_8P0D, inst);
 	dp_gswip_color_7P1D_set(GSW_MARKING_PCP_7P1D, inst);
@@ -991,8 +991,6 @@ static int dp_platform_color_table_set(int inst)
 	dp_gswip_remark_6P2D_set(GSW_MARKING_PCP_6P2D, inst);
 	dp_gswip_remark_5P3D_set(GSW_MARKING_PCP_5P3D, inst);
 	dp_gswip_remark_dscp_set(GSW_MARKING_DSCP_AF, inst);
-
-	return 0;
 }
 
 int dp_platform_queue_set(int inst, u32 flag)
@@ -1063,7 +1061,8 @@ int dp_platform_queue_set(int inst, u32 flag)
 		struct dp_subif_info *sif;
 
 		tx_push = &cpu_data.dq_tx_push_info[i];
-		if (tx_push->deq_port == (u32) - 1)
+		if (tx_push->deq_port < 0 ||
+		    tx_push->deq_port >= DP_MAX_CQM_DEQ)
 			continue;
 
 		DP_DEBUG(DP_DBG_FLAG_QOS, "cpu(%d) deq_port=%d",
@@ -1216,11 +1215,8 @@ static int dp_platform_set(int inst, u32 flag)
 
 			return -1;
 		}
-		if (dp_platform_color_table_set(inst)) {
-			pr_err("dp_platform_color_table_set fail\n");
+		dp_platform_color_table_set(inst);
 
-			return -1;
-		}
 		if (cpu_vlan_mod_dis(inst)) {
 			pr_err("cpu_vlan_mod_dis fail\n");
 			return -1;
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4.h b/drivers/net/datapath/dpm/gswip31/datapath_ppv4.h
index 312098c6cfa4..2dd52eaa8fa6 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4.h
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4.h
@@ -201,7 +201,7 @@ int dp_qos_port_conf_get(struct pp_qos_dev *qdev, u32 id,
 extern int (*qos_port_info_get)(struct pp_qos_dev *qdev, unsigned int id,
 				struct pp_qos_port_info *info);
 struct pp_qos_dev *dp_qos_dev_open(u32 id);
-int dp_map_to_drop_q(int inst, int q_id, struct dp_lookup_entry *lookup);
+void dp_map_to_drop_q(int inst, int q_id, struct dp_lookup_entry *lookup);
 int dp_pp_alloc_port(struct ppv4_port *info);
 int dp_pp_alloc_sched(struct ppv4_scheduler *info);
 int dp_pp_alloc_queue(struct ppv4_queue *info);
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
index 917174c69ee4..b9a236fe5285 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
@@ -356,11 +356,8 @@ static int queue_flush_31(int inst, int node_id, int flag)
 		goto EXIT;
 	}
 	/* map to drop queue and save the changed lookup entries for recover */
-	if (dp_map_to_drop_q(inst, qid, lookup)) {
-		pr_err("failed to dp_map_to_drop_q for Q:%d\n", qid);
-		res = DP_FAILURE;
-		goto EXIT;
-	}
+	dp_map_to_drop_q(inst, qid, lookup);
+
 	/* block/disable: ensure to drop all coming enqueue packet */
 	if (blocked == 0) { /* to block */
 		if (pp_qos_queue_block(priv->qdev, node_id)) {
@@ -1363,13 +1360,9 @@ int dp_node_alloc_31(struct dp_node_alloc *node, int flag)
 }
 
 /* dp_map_to_drop_q API
- * check index in lookup table for q_id
- * param q_buf: to save the lookup entries which mapped to this q_id
- * param num: q_buf size
- * set drop_q and return DP_SUCCESS
- * else return DP_FAILURE
+ * check index in lookup table for q_id and set drop_q
  */
-int dp_map_to_drop_q(int inst, int q_id, struct dp_lookup_entry *lookup)
+void dp_map_to_drop_q(int inst, int q_id, struct dp_lookup_entry *lookup)
 {
 	u32 i, k = 0;
 	struct hal_priv *priv = HAL(inst);
@@ -1385,7 +1378,6 @@ int dp_map_to_drop_q(int inst, int q_id, struct dp_lookup_entry *lookup)
 	}
 	if (lookup)
 		lookup->num = k;
-	return DP_SUCCESS;
 }
 
 /* dp_smart_free_from_child_31 API
@@ -3262,8 +3254,7 @@ int dp_node_link_en_set_31(struct dp_node_link_enable *en, int flag)
 	}
 
 	if (en->type == DP_NODE_QUEUE) {
-		if (!(en->en & (DP_NODE_EN | DP_NODE_DIS | DP_NODE_SUSPEND |
-				DP_NODE_RESUME))) {
+		if (en->en >= DP_NODE_SET_CMD_MAX) {
 			pr_err("Incorrect commands provided!\n");
 			return DP_FAILURE;
 		}
@@ -3310,8 +3301,7 @@ int dp_node_link_en_set_31(struct dp_node_link_enable *en, int flag)
 			return DP_FAILURE;
 		}
 	} else if (en->type == DP_NODE_PORT) {
-		if (!(en->en & (DP_NODE_EN | DP_NODE_DIS | DP_NODE_SUSPEND |
-				DP_NODE_RESUME))) {
+		if (en->en >= DP_NODE_SET_CMD_MAX) {
 			pr_err("Incorrect commands provided!\n");
 			return DP_FAILURE;
 		}
@@ -3687,8 +3677,9 @@ int dp_shaper_conf_set_31(struct dp_shaper_conf *cfg, int flag)
 			return DP_FAILURE;
 		}
 
-		if (cfg->cmd == DP_SHAPER_CMD_ADD ||
-		    cfg->cmd == DP_SHAPER_CMD_ENABLE) {
+		switch(cfg->cmd) {
+		case DP_SHAPER_CMD_ADD:
+		case DP_SHAPER_CMD_ENABLE:
 			res = limit_dp2pp(cfg->cir, &bw_limit);
 
 			if (res == DP_FAILURE) {
@@ -3696,10 +3687,12 @@ int dp_shaper_conf_set_31(struct dp_shaper_conf *cfg, int flag)
 				return DP_FAILURE;
 			}
 			queue_cfg.common_prop.bandwidth_limit = bw_limit;
-		} else if (cfg->cmd == DP_SHAPER_CMD_REMOVE ||
-			   cfg->cmd == DP_SHAPER_CMD_DISABLE) {
+			break;
+		case DP_SHAPER_CMD_REMOVE:
+		case DP_SHAPER_CMD_DISABLE:
 			queue_cfg.common_prop.bandwidth_limit = 0;
-		} else {
+			break;
+		default:
 			pr_err("Incorrect command provided:%d\n", cfg->cmd);
 			return DP_FAILURE;
 		}
@@ -3727,8 +3720,9 @@ int dp_shaper_conf_set_31(struct dp_shaper_conf *cfg, int flag)
 			return DP_FAILURE;
 		}
 
-		if (cfg->cmd == DP_SHAPER_CMD_ADD ||
-		    cfg->cmd == DP_SHAPER_CMD_ENABLE) {
+		switch(cfg->cmd) {
+		case DP_SHAPER_CMD_ADD:
+		case DP_SHAPER_CMD_ENABLE:
 			res = limit_dp2pp(cfg->cir, &bw_limit);
 
 			if (res == DP_FAILURE) {
@@ -3736,10 +3730,12 @@ int dp_shaper_conf_set_31(struct dp_shaper_conf *cfg, int flag)
 				return DP_FAILURE;
 			}
 			sched_cfg.common_prop.bandwidth_limit = bw_limit;
-		} else if (cfg->cmd == DP_SHAPER_CMD_REMOVE ||
-			   cfg->cmd == DP_SHAPER_CMD_DISABLE) {
+			break;
+		case DP_SHAPER_CMD_REMOVE:
+		case DP_SHAPER_CMD_DISABLE:
 			sched_cfg.common_prop.bandwidth_limit = 0;
-		} else {
+			break;
+		default:
 			pr_err("Incorrect command provided:%d\n", cfg->cmd);
 			return DP_FAILURE;
 		}
@@ -3769,8 +3765,9 @@ int dp_shaper_conf_set_31(struct dp_shaper_conf *cfg, int flag)
 			return DP_FAILURE;
 		}
 
-		if (cfg->cmd == DP_SHAPER_CMD_ADD ||
-		    cfg->cmd == DP_SHAPER_CMD_ENABLE) {
+		switch(cfg->cmd) {
+		case DP_SHAPER_CMD_ADD:
+		case DP_SHAPER_CMD_ENABLE:
 			res = limit_dp2pp(cfg->cir, &bw_limit);
 
 			if (res == DP_FAILURE) {
@@ -3778,10 +3775,12 @@ int dp_shaper_conf_set_31(struct dp_shaper_conf *cfg, int flag)
 				return DP_FAILURE;
 			}
 			port_cfg.common_prop.bandwidth_limit = bw_limit;
-		} else if (cfg->cmd == DP_SHAPER_CMD_REMOVE ||
-			   cfg->cmd == DP_SHAPER_CMD_DISABLE) {
+			break;
+		case DP_SHAPER_CMD_REMOVE:
+		case DP_SHAPER_CMD_DISABLE:
 			port_cfg.common_prop.bandwidth_limit = 0;
-		} else {
+			break;
+		default:
 			pr_err("Incorrect command provided:%d\n", cfg->cmd);
 			return DP_FAILURE;
 		}
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_proc.c b/drivers/net/datapath/dpm/gswip31/datapath_proc.c
index 8dfc6a1cc37a..2d5d1ada0acc 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_proc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_proc.c
@@ -322,7 +322,7 @@ int dp_sub_proc_install_31(void)
 #define INVALID_DMA_CH 255
 char *get_dma_flags_str31(u32 epn, char *buf, int buf_len)
 {
-	char tmp[30]; /*must be static */
+	char tmp[40];
 	u32 flags;
 	u32 tx_ch, k;
 	u8 f_found;
@@ -341,6 +341,7 @@ char *get_dma_flags_str31(u32 epn, char *buf, int buf_len)
 		if (port->flag_other & CBM_PORT_DMA_CHAN_SET &&
 		    port->deq_port_base == epn) {
 			tx_ch = port->dma_chan;
+			flags = port->alloc_flags;
 			break;
 		}
 	}
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_switchdev.c b/drivers/net/datapath/dpm/gswip31/datapath_switchdev.c
index dbd15d200ad4..912911866771 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_switchdev.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_switchdev.c
@@ -21,7 +21,7 @@ int dp_swdev_alloc_bridge_id(int inst)
 	gsw_handle = dp_port_prop[inst].ops[0];
 	memset(&br, 0, sizeof(br));
 	ret = gsw_handle->gsw_brdg_ops.Bridge_Alloc(gsw_handle, &br);
-	if (ret != GSW_statusOk || br.nBridgeId < 0) {
+	if (ret != GSW_statusOk) {
 		pr_err("Failed to get a FID\n");
 		return -1;
 	}
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_tx.c b/drivers/net/datapath/dpm/gswip31/datapath_tx.c
index aa0df7797e9d..bec1274072a2 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_tx.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_tx.c
@@ -409,39 +409,14 @@ int32_t dp_xmit_31(struct net_device *rx_if, dp_subif_t *rx_subif,
 
 lbl_err_ret:
 	switch (err_ret) {
-	case DP_XMIT_ERR_NOT_INIT:
-		printk_ratelimited("dp_xmit failed for dp no init yet\n");
-		break;
-	case DP_XMIT_ERR_IN_IRQ:
-		printk_ratelimited("dp_xmit not allowed in interrupt context\n");
-		break;
-	case DP_XMIT_ERR_NULL_SUBIF:
-		printk_ratelimited("dp_xmit failed for rx_subif null\n");
-		UP_STATS(get_dp_port_info(inst, 0)->tx_err_drop);
-		break;
 	case DP_XMIT_ERR_PORT_TOO_BIG:
 		UP_STATS(get_dp_port_info(inst, 0)->tx_err_drop);
 		printk_ratelimited("rx_subif->port_id >= max_ports");
 		break;
-	case DP_XMIT_ERR_NULL_SKB:
-		printk_ratelimited("skb NULL");
-		UP_STATS(get_dp_port_info(inst,
-					  rx_subif->port_id)->tx_err_drop);
-		break;
 	case DP_XMIT_ERR_NULL_IF:
 		UP_STATS(mib->tx_pkt_dropped);
 		printk_ratelimited("rx_if NULL");
 		break;
-	case DP_XMIT_ERR_REALLOC_SKB:
-		pr_info_once("dp_create_new_skb failed\n");
-		break;
-	case DP_XMIT_ERR_EP_ZERO:
-		pr_err("Why ep zero in dp_xmit for %s\n",
-		       skb->dev ? skb->dev->name : "NULL");
-		break;
-	case DP_XMIT_ERR_GSO_NOHEADROOM:
-		pr_err("No enough skb headerroom(GSO). Need tune SKB buffer\n");
-		break;
 	case DP_XMIT_ERR_CSM_NO_SUPPORT:
 		printk_ratelimited("dp_xmit not support checksum\n");
 		break;
diff --git a/include/net/datapath_api_qos.h b/include/net/datapath_api_qos.h
index 35ccfcf3e3af..1e97fc79aa91 100644
--- a/include/net/datapath_api_qos.h
+++ b/include/net/datapath_api_qos.h
@@ -374,10 +374,11 @@ enum dp_node_en {
 				   *   Not for TMU
 				   *   for ppv4, valid queue/sched/port
 				   */
-	DP_NODE_RESUME = BIT(3)   /*!< Resume scheduling for this node:
+	DP_NODE_RESUME = BIT(3),   /*!< Resume scheduling for this node:
 				   *  Not for TMU
 				   *  for ppv4,valid queue/sched/port
 				   */
+	DP_NODE_SET_CMD_MAX = BIT(4)
 };
 
 /*! @brief arbitration method used for the node in its parents scheduler/ports*/
